Fix URICheckEncodingMonitor NPE

MarkEWaite opened the pull request:
## Fix URICheckEncodingMonitor null pointer exception

A [null pointer exception report](https://community.jenkins.io/t/reverse-proxy-test-does-not-handle-url-correctly/1294/9) on https://community.jenkins.io/ notes that if the `value` returned by `URICheckEncodingMonitor` is the empty string or does not exist, then a null pointer exception is throws.  Since the `URICheckEncodingMonitor` is checking for configuration errors and one of the (rather obscure) configuration errors might be to return an empty string, this pull request provides a test that checks the failure case and prevents the failure with a null check.

No Jira ticket was opened for the issue.  I can open a Jira ticket if that is required.

Test fails without the change and passes with the change.

- Test null pointer exception in URICheckEncodingMonitor
- Avoid URICheckEncodingMonitor empty value NPE

### Testing done

Automated test written to confirm the failure is visible.  Automated test run after the fix to confirm it passes.  Interactive test run to start Jenkins and confirm that the monitor causes no unexpected side effects.

### Proposed changelog entries

- No changelog - too minor

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7543"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

MarkEWaite committed:
Test null pointer exception in URICheckEncodingMonitor

When value in the stapler request is empty or does not exist, a null
pointer exception is reported.

MarkEWaite committed:
Avoid URICheckEncodingMonitor empty value NPE

https://community.jenkins.io/t/reverse-proxy-test-does-not-handle-url-correctly/1294/9
reports that an empty value returned to the URI encoding check monitor
will report a null pointer exception.  An empty value is unexpected in
that context, but it is easy to check for null and avoid the null
pointer exception.

No Jira ticket was opened for the issue.  I can open a Jira ticket if
that is required.

basil commented on core/src/main/java/jenkins/diagnostics/URICheckEncodingMonitor.java:
```diff
@@ -43,7 +43,7 @@ public FormValidation doCheckURIEncoding(StaplerRequest request) throws IOExcept
         if (!expected.equals(value)) {
             String expectedHex = Util.toHexString(expected.getBytes(StandardCharsets.UTF_8));
-            String valueHex = Util.toHexString(value.getBytes(StandardCharsets.UTF_8));
+            String valueHex = value != null ? Util.toHexString(value.getBytes(StandardCharsets.UTF_8)) : "<empty string>";
```
(Optional suggestion.) Seems inconsistent to me that when the "value" parameter is empty that `value` would be `null` but `valueHex` would be "empty string". The below is more consistent but it does not matter either way.

```suggestion
            String valueHex = value != null ? Util.toHexString(value.getBytes(StandardCharsets.UTF_8)) : null;
```

MarkEWaite committed:
Set hex value to null if value is null

Retain internal consistency in the code.

Since value is set to null if it is an empty string,
it makes sense to set valueHex to null if value
is null.

Co-authored-by: Basil Crow <me@basilcrow.com>

MarkEWaite commented:
This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.
